### PR TITLE
Boolean QA framework for `run_mrc.py` 

### PR DESCRIPTION
**Pull request by jsmcibm**
Results on the dev set with command
```

python examples/mrc/run_mrc.py --model_name_or_path ${MRCMODEL} \
       --output_dir /dccstor/jsmc-nmt-01/bool/expts/toolkit/b/b29a/ \
       --fp16 --per_device_eval_batch_size 128 --overwrite_output_dir \
       --do_boolean --boolean_config examples/boolqa/tydi_boolqa_config.json \
       --overwrite_cache
       ```
was
```
{
    "avg_minimal_f1": 0.7161550290871469,
    "avg_minimal_precision": 0.7546935029120848,
    "avg_minimal_recall": 0.6840807371529009,
    "avg_passage_f1": 0.7482995393781143,
    "avg_passage_precision": 0.7762910462852008,
    "avg_passage_recall": 0.7235303869218735
}
```

### CODE REVIEW

File: tests/primeqa/boolqa/processors/preprocessors/test_boolqa_classifer_preprocessor.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    _eval_pred_in_file = 'tests/resources/boolqa/processors/preprocessors/eval_predictions.json'
+    _expected_eval_dataset0 = {'example_id': '9d854612-d484-4b98-b208-4983baeeab67',
+                              'language': 'english',
+                              'question': 'Do zebra finches have stripes?',
+                              'label': 0,
+                              'span_answer_text': 'There are two subspecies of the zebra finch:',
+                              'input_ids': [0, 984, 116232, 2270, 17007, 765, 43613, 90, 32, 2, 2, 8622, 621, 6626, 1614, 16711, 3387, 111, 70, 116232, 2270, 206, 12, 2],
+                              'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]
+                              }
+    _evc_model_path='/dccstor/jsmc-nmt-01/bool/git/IOTA-boolean-challenge/model/evc-c5'
```
Comment (bhavani105) [inline tests/primeqa/boolqa/processors/preprocessors/test_boolqa_classifer_preprocessor.py:19]: fix path

**Review by bhavani105 — COMMENTED** [review]
Looks good. Left some comments

File: primeqa/boolqa/processors/dataset/mrc2dataset.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+                val['rank']=rank
+                yield val
+    df=pd.DataFrame.from_records(read_mrc_inner())
+    df0=df.query('rank==0') # TODO danger - assumes upstream results are correctly ordered
+    ds=Dataset.from_pandas(df0, preserve_index=False)
+    return ds
+
+#----------------------------------------------------------------
+def main():
+#    args=handle_args()
```
[the remark this answers is not in the document]
Comment (jsmcibm) [reply primeqa/boolqa/processors/dataset/mrc2dataset.py]: main is not being used so I removed it

File: primeqa/boolqa/processors/preprocessors/boolqa_classifier.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+        if examples.num_rows == 0:
+            raise ValueError("No examples to process")
+        if not 'question' in examples.column_names:
+            msg="""
+The eval_predictions.json file must contain a field 'question'.  This can be created by running the machine
+reading comprehension step in "run_mrc.py" with the option
+--postprocessor primeqa.boolqa.processors.postprocessors.extractive.ExtractivePipelinePostProcessor
+rather than the default tydi postprocessor.
+            """
+            print(msg)
```
[the remark this answers is not in the document]
Comment (jsmcibm) [reply primeqa/boolqa/processors/preprocessors/boolqa_classifier.py]: done

File: tests/primeqa/boolqa/processors/preprocessors/test_boolqa_classifer_preprocessor.py (hunk 1 above)
Comment (jsmcibm) [reply tests/primeqa/boolqa/processors/preprocessors/test_boolqa_classifer_preprocessor.py]: done

**Comment by avisil** [comment]
I changed the name of the PR to be a bit more meaningful: Boolean QA framework for `run_mrc.py`

File: oneqa/boolqa/processors/postprocessors/extractive.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
+from oneqa.mrc.processors.postprocessors.extractive import ExtractivePostProcessor
+from oneqa.mrc.data_models.eval_prediction_with_processing import EvalPredictionWithProcessing
+
+from datasets import Dataset
+from tqdm import tqdm
+
+from typing import List, Dict, Any, Tuple
+
+class ExtractivePipelinePostProcessor(ExtractivePostProcessor):
+
```
Comment (rongzhangibm) [inline oneqa/boolqa/processors/postprocessors/extractive.py]: Suggest to add doc string to this class and other classes in pre/post processors.

**Review by rongzhangibm — APPROVED** [review]
LGTM.

File: examples/mrc/run_mrc.py [hunk 5]
```diff
@@ -440,6 +465,38 @@ def compute_metrics(p: EvalPredictionWithProcessing):
...
+        sn_main([boolean_config['sn']])
+
+        with open(os.path.join(boolean_config['sn']['output_dir'], 'eval_predictions_processed.json'), 'r') as f:
+            processed_predictions = json.load(f)
+        references = postprocessor.prepare_examples_as_references(eval_examples)
+        boolean_eval_metric = eval_metrics.compute(predictions=processed_predictions, references=references)
+        path = os.path.join(boolean_config['sn']['output_dir'], f"all_results.json")
+        with open(path, "w") as f:
+            json.dump(boolean_eval_metric, f, indent=4, sort_keys=True)        
+
```
Comment (rongzhangibm) [inline examples/mrc/run_mrc.py:499]: Could we move the do_boolean codes into a function i.e. do_boolean(data_args, training_args, model_args, task_config, ...) and let main() call the function?

**Review by rongzhangibm — APPROVED** [review]
LGTM

File: examples/mrc/run_mrc.py (hunk 5 above)
Comment (jsmcibm) [reply examples/mrc/run_mrc.py:499]: I don't think its necessary yet, but if it gets any bigger I agree.  Probably there should be some coordination with other functionalities in run_mrc.py

File: tests/primeqa/boolqa/processors/postprocessors/test_extractive_pipeline_postprocessor.py [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+            predicted = preds[0]
+            expected = mock_logits_expected_predictions['expected_predictions'][example_id]
+            expected_start_end = expected_start_end_index[example_id]
+            ptargettype = int(np.argmax(predicted['target_type_logits']))
+            assert (predicted['question'] == expected['question'] )
+            assert (predicted['language'] == expected['language'] )
+            assert (predicted['start_index'], predicted['end_index']) == expected_start_end
+            assert predicted['passage_index']  == expected['passage_index'] 
+            assert ptargettype  == expected['target_type'] 
+
```
Comment (bhavani105) [inline tests/primeqa/boolqa/processors/postprocessors/test_extractive_pipeline_postprocessor.py:152]: Maybe take a look at the SQUAD postprocessor test where it only checks the additional fields
https://github.com/primeqa/primeqa/blob/c6cf549bcc2e0e4fde1cfbfa31aa8a75c73c06be/tests/primeqa/mrc/processors/postprocessors/test_squad_postprocessor.py#L94

File: tests/primeqa/boolqa/processors/preprocessors/test_boolqa_classifer_preprocessor.py [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+                              'language': 'english',
+                              'question': 'Do zebra finches have stripes?',
+                              'label': 0,
+                              'span_answer_text': 'There are two subspecies of the zebra finch:',
+                              'input_ids': [0, 984, 116232, 2270, 17007, 765, 43613, 90, 32, 2, 2, 8622, 621, 6626, 1614, 16711, 3387, 111, 70, 116232, 2270, 206, 12, 2],
+                              'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]
+                              }
+    _evc_model_path='xlm-roberta-large'
+    
+
```
Comment (bhavani105) [inline tests/primeqa/boolqa/processors/preprocessors/test_boolqa_classifer_preprocessor.py:21]: The obove should preferably be fixtures

**Review by bhavani105 — APPROVED** [review]
LGTM

File: tests/primeqa/boolqa/processors/postprocessors/test_extractive_pipeline_postprocessor.py (hunk 6 above)
Comment (jsmcibm) [reply tests/primeqa/boolqa/processors/postprocessors/test_extractive_pipeline_postprocessor.py:152]: it seems safer to me to have a few extra asserts in the unit tests

Comment (jsmcibm) [reply tests/primeqa/boolqa/processors/postprocessors/test_extractive_pipeline_postprocessor.py:152]: it seems safer to me to have a few extra asserts for the unit tests

**Review by avisil — APPROVED** [review]
The Travis tests have passed - congrats! I approve this PR.

**Pull request merged**